Generate serialization for NSError

sheeparegreat opened the pull request:
#### 748284bf1db037bd7ed18324fb23de20605c06af
<pre>
Generate serialization for NSError
<a href="https://bugs.webkit.org/show_bug.cgi?id=264462">https://bugs.webkit.org/show_bug.cgi?id=264462</a>
<a href="https://rdar.apple.com/118155518">rdar://118155518</a>

Reviewed by Brady Eidson.

* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.h:
* Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.mm:
(IPC::typeFromObject):
* Source/WebKit/Shared/Cocoa/CoreIPCError.h: Copied from Source/WebKit/Shared/Cocoa/CoreIPCNSCFObject.h.
(WebKit::CoreIPCError::CoreIPCError):
(WebKit::CoreIPCError::domain const):
(WebKit::CoreIPCError::code const):
(WebKit::CoreIPCError::userInfo const):
(WebKit::CoreIPCError::underlyingError const):
* Source/WebKit/Shared/Cocoa/CoreIPCError.mm: Added.
(WebKit::CoreIPCError::hasValidUserInfo):
(WebKit::CoreIPCError::toID const):
(WebKit::CoreIPCError::isSafeToEncodeUserInfo):
(WebKit::CoreIPCError::CoreIPCError):
* Source/WebKit/Shared/Cocoa/CoreIPCError.serialization.in: Added.
* Source/WebKit/Shared/Cocoa/CoreIPCNSCFObject.h:
* Source/WebKit/Shared/Cocoa/CoreIPCNSCFObject.mm:
(WebKit::valueFromID):
* Source/WebKit/Shared/mac/WebCoreArgumentCodersMac.mm:
(IPC::ArgumentCoder&lt;WebCore::ResourceError&gt;::encodePlatformData):
(IPC::ArgumentCoder&lt;WebCore::ResourceError&gt;::decodePlatformData):
(IPC::isSafeToEncodeUserInfo): Deleted.
(IPC::encodeNSError): Deleted.
(IPC::decodeNSError): Deleted.
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/IPC/IPCSerialization.mm:
(createCertificate):
(createPrivateKey):
(TEST):

Canonical link: <a href="https://commits.webkit.org/270706@main">https://commits.webkit.org/270706@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1bf4d29acca8a5c8e0d04c686250d7e6faa911e5

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/26037 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/4649 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/27319 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/28135 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/23847 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/26358 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/6414 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/2077 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/23912 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/26291 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/3527 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/22435 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/28716 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/3147 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/23387 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/29445 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/23763 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/23767 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/27326 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/3177 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/1385 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/4569 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6294 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/3637 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/3498 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

beidson commented on Source/WebKit/SourcesCocoa.txt:
```diff
@@ -136,6 +136,7 @@ Shared/WebSQLiteDatabaseTracker.cpp
 
 Shared/cf/ArgumentCodersCF.cpp @no-unify
 Shared/cf/CookieStorageUtilsCF.mm
+Shared/cf/CoreIPCError.mm
```
This should really be in Shared/Cocoa/ - It's doing NSError, not CFError.

beidson reviewed (approved):
Shared/cocoa seems much more appropriate for this one.

Looks good otherwise.

achristensen07 commented on Source/WebKit/Shared/cf/CoreIPCError.h:
```diff
@@ -0,0 +1,85 @@
+    CoreIPCError& operator=(CoreIPCError&&) = default;
+
+    CoreIPCError(NSError *);
+    CoreIPCError(String&& domain, int64_t&& code, RetainPtr<CFDictionaryRef>&& userInfo, std::unique_ptr<CoreIPCError>&& underlyingError)
```
int64_t doesn't need &&

achristensen07 commented on Source/WebKit/Shared/cf/CoreIPCError.h:
```diff
@@ -0,0 +1,85 @@
+
+#pragma once
+
+#if USE(CF)
```
PLATFORM(COCOA), but at this point they mean basically the same thing.  There used to be a Windows port that used CoreFoundation but not Foundation.

achristensen07 commented on Tools/TestWebKitAPI/Tests/IPC/IPCSerialization.mm:
```diff
@@ -288,4 +289,14 @@ id valueAsID() const
+    NSError *error1 = [NSError errorWithDomain:@"TestWTFDomain" code:1 userInfo:@{ NSLocalizedDescriptionKey : @"TestWTF Error" }];
+    NSError *error2 = [NSError errorWithDomain:@"TestWTFDomain" code:2 userInfo:@{
+        NSLocalizedDescriptionKey: @"Localized Description",
+        NSLocalizedFailureReasonErrorKey: @"Localized Failure Reason Error",
```
It would be good to add an array, a dictionary, a URL, a number, a cert chain and a SecTrustRef with key NSErrorPeerCertificateChainKey, an identity with key NSErrorClientCertificateChainKey, and an unexpected value like a number with key NSErrorClientCertificateChainKey.  We already have functions called testIdentity(), testCertificate(), and secTrustFromCertificateChain that may need some moving to be used here.

sheeparegreat committed:
Generate serialization for NSError
https://bugs.webkit.org/show_bug.cgi?id=264462
rdar://118155518

Reviewed by Brady Eidson.

* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.h:
* Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.mm:
(IPC::typeFromObject):
* Source/WebKit/Shared/Cocoa/CoreIPCError.h: Copied from Source/WebKit/Shared/Cocoa/CoreIPCNSCFObject.h.
(WebKit::CoreIPCError::CoreIPCError):
(WebKit::CoreIPCError::domain const):
(WebKit::CoreIPCError::code const):
(WebKit::CoreIPCError::userInfo const):
(WebKit::CoreIPCError::underlyingError const):
* Source/WebKit/Shared/Cocoa/CoreIPCError.mm: Added.
(WebKit::CoreIPCError::hasValidUserInfo):
(WebKit::CoreIPCError::toID const):
(WebKit::CoreIPCError::isSafeToEncodeUserInfo):
(WebKit::CoreIPCError::CoreIPCError):
* Source/WebKit/Shared/Cocoa/CoreIPCError.serialization.in: Added.
* Source/WebKit/Shared/Cocoa/CoreIPCNSCFObject.h:
* Source/WebKit/Shared/Cocoa/CoreIPCNSCFObject.mm:
(WebKit::valueFromID):
* Source/WebKit/Shared/mac/WebCoreArgumentCodersMac.mm:
(IPC::ArgumentCoder<WebCore::ResourceError>::encodePlatformData):
(IPC::ArgumentCoder<WebCore::ResourceError>::decodePlatformData):
(IPC::isSafeToEncodeUserInfo): Deleted.
(IPC::encodeNSError): Deleted.
(IPC::decodeNSError): Deleted.
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/IPC/IPCSerialization.mm:
(createCertificate):
(createPrivateKey):
(TEST):

Canonical link: https://commits.webkit.org/270706@main
